File: galois-linear-feedback-shift-register/LfsrRandom.java
/* 
 * Galois linear feedback shift register (LFSR) (Java)
 * 
 * Copyright (c) 2021 Project Nayuki
 * All rights reserved. Contact Nayuki for licensing.
 * https://www.nayuki.io/page/galois-linear-feedback-shift-register
 */

import java.math.BigInteger;
import java.util.Random;


/**
 * Galois LFSR random number generator. This can be used in place of java.util.Random.
 * In this class, a polynomial is represented as a BigInteger, where the coefficient of
 * the <var>x</var><sup><var>k</var></sup> term is represented by bit <var>k</var>.
 */
public final class LfsrRandom extends Random {
	
	/*---- Demo program ----*/
	
	public static void main(String[] args) {
		LfsrRandom r = new LfsrRandom(BigInteger.valueOf(0x16801), BigInteger.valueOf(1));
		r.printDebug();
		for (int i = 0; i < 30; i++)
			System.out.printf("%08x%n", r.nextInt());
		r.printDebug();
	}
	
	
	
	/*---- Instance members ----*/
	
	private final BigInteger characteristic;
	private final int degree;
	
	private BigInteger state;
	
	
	/**
	 * Constructs an LFSR with the specified characteristic polynomial and initial state polynomial.
	 * The characteristic polynomial must have degree at least 2. The state polynomial must have degree
	 * less than the degree of the characteristic polynomial, and must not be the zero polynomial.
	 * @param charistic the characteristic polynomial
	 * @param state the initial state polynomial
	 * @throws NullPointerException if any argument is {@code null}
	 * @throws IllegalArgumentException if the polynomials do not satisfy the requirements
	 */
	public LfsrRandom(BigInteger charis, BigInteger state) {
		if (charis.signum() == -1)
			throw new IllegalArgumentException("Invalid characteristic polynomial - negative");
		if (charis.bitLength() < 2)
			throw new IllegalArgumentException("Invalid characteristic polynomial - degree too low");
		if (state.equals(BigInteger.ZERO))
			throw new IllegalArgumentException("Invalid state polynomial - all zero");
		if (state.bitLength() >= charis.bitLength())
			throw new IllegalArgumentException("Invalid state polynomial - degree >= char poly degree");
		
		characteristic = charis;
		degree = charis.bitLength() - 1;
		this.state = state;
	}
	
	
	@Override public boolean nextBoolean() {
		boolean result = state.testBit(0);      // Use bit 0 in the LFSR state as the result
		state = state.shiftLeft(1);             // Multiply by x
		if (state.testBit(degree))              // If degree of state polynomial matches degree of characteristic polynomial
			state = state.xor(characteristic);  // Then subtract the characteristic polynomial from the state polynomial
		return result;
	}
	
	
	@Override protected int next(int bits) {
		int result = 0;
		for (int i = 0; i < bits; i++)
			result = (result << 1) | (nextBoolean() ? 1 : 0);
		return result;
	}
	
	
	public void printDebug() {
		System.out.printf("characteristic: degree=%d  poly = %s%nstate: degree=%d  poly = %s%n",
			degree, polynomialToString(characteristic), state.bitLength() - 1, polynomialToString(state));
	}
	
	
	private static String polynomialToString(BigInteger poly) {
		StringBuilder sb = new StringBuilder();
		boolean head = true;
		for (int i = poly.bitLength() - 1; i >= 0; i--) {
			if (poly.testBit(i)) {
				if (head) head = false;
				else sb.append(" + ");
				sb.append("x^" + i);
			}
		}
		return sb.toString();
	}
	
}

File: galois-linear-feedback-shift-register/lfsrrandom.py
# 
# Galois linear feedback shift register (LFSR) (Python)
# 
# Copyright (c) 2021 Project Nayuki
# All rights reserved. Contact Nayuki for licensing.
# https://www.nayuki.io/page/galois-linear-feedback-shift-register
# 

import random


# Random number generator class (implements most functionality of random.Random)
class LfsrRandom(random.Random):
	
	def __new__(cls, *args, **kwargs):  # Magic because the superclass doesn't cooperate
		return random.Random.__new__(cls, random.random())
	
	
	def __init__(self, charis: int, state: int):
		assert isinstance(charis, int)
		assert isinstance(state, int)
		
		if charis < 0:
			raise ValueError("Invalid characteristic polynomial - negative")
		if charis.bit_length() < 2:
			raise ValueError("Invalid characteristic polynomial - degree too low")
		if state == 0:
			raise ValueError("Invalid state polynomial - all zero")
		if state.bit_length() >= charis.bit_length():
			raise ValueError("Invalid state polynomial - degree >= char poly degree")
		
		self.characteristic = charis
		self.degree = charis.bit_length() - 1
		self.state = state
	
	
	def randbit(self) -> int:
		result: int = self.state & 1                   # Use bit 0 in the LFSR state as the result
		self.state = self.state << 1              # Multiply by x
		if (self.state >> self.degree) & 1 != 0:  # If degree of state polynomial matches degree of characteristic polynomial
			self.state ^= self.characteristic     # Then subtract the characteristic polynomial from the state polynomial
		return result
	
	
	def getrandbits(self, k: int) -> int:
		result: int = 0
		for i in range(k):
			result = (result << 1) | self.randbit()
		return result
	
	
	def random(self) -> float:
		return self.getrandbits(52) / (1 << 52)



# Demo main program
if __name__ == "__main__":
	# Polynomial: x^16 + x^14 + x^13 + x^11 + x^0
	rand = LfsrRandom(0b10110100000000001, 1)
	for i in range(10):
		print(rand.random())
	for i in range(20):
		print(rand.randbit())
